ceph_salt/config_shell.py

```diff
@@ -663,6 +705,97 @@ def ui_command_remove(self, value):
+        for idx, item in enumerate(items):
+            ListElementNode(str(idx), self, item)
+
+    def _remove_all_children(self):
+        for idx in range(len(self.value)):
+            self.remove_child(self.get_child(str(idx)))
+
+    def _list_commands(self):
+        return ['add', 'remove']
+
+    def summary(self):
+        value_list, _ = self._find_value()
+        return str(len(value_list)) if value_list else 'empty', None
+
+    def _precheck_params(self, spec, kwargs):
+        # check required parameters
+        all_params = set(spec.keys())
+        required_params = {k for k, v in spec.items() if v.get('required', False)}
+        assert required_params  # at least one parameter is required
+        missing_params = required_params - set(kwargs.keys())
+        if missing_params:
+            raise Exception('Required parameter(s) `{}` are missing.'.format(
+                ', '.join(missing_params)))
+        # check if there are any unknown parameters
+        unknown_params = set(kwargs.keys()) - all_params
+        if unknown_params:
+            raise Exception('Unknown parameter(s): `{}`.'.format(', '.join(unknown_params)))
+
+    def _format_params(self, kwargs):
+        """Validate and format input parameters."""
+        spec = self.option_dict.get('params_spec')
+        self._precheck_params(spec, kwargs)
+        value = {}
+        for k, v in kwargs.items():
+            # validate the input value
+            validator = spec[k].get('validator')
+            if validator and not validator.validate(v):
+                raise Exception("Invalid value {} for {}".format(v, k))
+            # transform to native type from str
+            transformer = spec[k].get('transformer')
+            value[k] = transformer.transform(v) if transformer else v
+        return value
+
+    def ui_command_add(self, **kwargs):
+        new_item = self._format_params(kwargs)
+        if new_item not in self.value:
+            ListElementNode(str(len(self.value)), self, new_item)
+            self.value.append(new_item)
+            self.option_dict['handler'].save(self.value)
+            PP.pl_green('Value added.')
+        else:
+            PP.pl_red('Value already exists.')
+
+    def ui_command_remove(self, **kwargs):
+        remove_item = self._format_params(kwargs)
+
+        # do not include items that have specified KVs
+        def __match(item):
+            for search_k, search_v in remove_item.items():
+                if item.get(search_k) != search_v:
+                    break
+            else:
+                return True
+            return False
+        new_value = list(itertools.filterfalse(__match, self.value))
+
+        if len(new_value) != len(self.value):
+            self._remove_all_children()
+            self.value = new_value
+            self._add_children(self.value)
+            self.option_dict['handler'].save(self.value)
+            PP.pl_green('Value removed.')
```

Comment:
nit: For consistency, and better usability, can we show the number of removed items? Something like https://github.com/ceph/ceph-salt/blob/master/ceph_salt/config_shell.py#L837-L840

Reply:
Hmm.. how about renaming the term to `items`? add item to a list, remove items from list. `Value` is a bit weird for dictionaries.

Reply:
Fixed, thanks!